Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -68,6 +69,27 @@ class Writer {
 
   void EncodeString(const std::string& string_value);
 
+  /**
+   * Encodes a message and its length.
+   *
+   * When encoding a top level message, protobuf doesn't include the length
+   * (since you can get that already from the length of the binary output.) But
+   * when encoding a sub/nested message, you must include the length in the
+   * serialization.
+   *
+   * Call this method when encoding a non top level message. Provide a function
+   * to encode the message itself. This method will calculate the size of the
+   * encoded message (using the provided function with a non-writing sizing
+   * stream), write out the size (and perform sanity checks), and then serialize
+   * the message by calling the provided function a second time.
+   */
+  void EncodeNestedMessage(
+      const std::function<void(Writer&)>& encode_message_fn);
```

Comment:
Alternative: Costa points out that rather than using std::function, we can just use a template callback, i.e.

```
template <typename CallbackT>
void EncodeNestedMessage(const CallbackT& encode_message_fn);
```

Although the signature of the callback is less obvious, it does save some size (and speed; though that's likely not as important here.) To benchmark at least the size, I compiled this both ways and looked at the size of the resulting .a file, as well as the size of the unittest binary for this class. Results:

firebase_firestore_remote_test:
std::function: 11M (11328504)
CallbackT: 11M (11304840)

libfirebase_firestore_remote.a:
std::function: 554K (556886)
CallbackT: 513K (524444)

(This is with our default set of flags, which do not seem to include an optimization flag.)

So the total damage to the resulting binary is 23664 bytes, which is surprisingly large (to me at least.)

Perhaps more problematic (though unrelated to this PR) is the size of the .a file regardless. Half a meg for just the serialization stuff is ... big. (Yes, there's more in remote/ than just serializer.cc, but not much more. serializer.cc.o is by far the dominant contributor to the size of the .a file.)

I'm going to investigate this size issue a bit more.

Reply:
Oh hold on; we've got debug symbols enabled. Let me play around with the flags a bit more.

Reply:
Ok; sanity is (mostly[1]) restored.

Compiling with `cmake -DCMAKE_BUILD_TYPE=MinSizeRel` should cause debug symbols to not be included (at least for our code). Using that I get the following:

firebase_firestore_remote_test:
- 6.7M (7013664)
- 6.7M (7011504)

libfirebase_firestore_remote.a:
- 29K (28816)
- 25K (24958)

`strip`ing the results, I get:

firebase_firestore_remote_test:
- 2.0M (2082536)
- 2.0M (2082504)

libfirebase_firestore_remote.a:
- 9.9K (10064)
- 9.0K (9192)

(much more reasonable.)

So back to the original question: The total damage done to the resulting binary is 32 bytes. I'm inclined to leave it as std::function for the extra readability.


[1] - curiously, some of our unit tests fail when compiled with MinSizeRel. :(  I'll investigate this and send out a few PRs.

Reply:
^ err, to be clear, the first number in each is the std::function number, and the second is the CallbackT number. (Just like my first comment.)

Reply:
Thanks for benchmarking this. Looks like this usage gets optimized very nicely.

Reply:
Given that we intend this function to be used by multiple callers, the savings we get due to the type erasure performed by std::function will overwhelm the template-based method. The mild performance hit due to std::function just isn't a concern.